ui/src/app/views/notebook-resource-card.tsx

```diff
@@ -0,0 +1,188 @@
+        },
+        {
+          displayName: 'Delete',
+          onClick: () => {
+            this.props.showConfirmDeleteModal(this.displayName, 'Notebook',
+              (onComplete) => this.deleteNotebook(onComplete));
+          },
+        },
+        {
+          displayName: 'Open in Jupyter Lab',
+          onClick: () => navigateByUrl(this.resourceUrl(true)),
+        }];
+    }
+
+    get readOnly(): boolean {
+      return this.props.resourceCard.permission === 'READER';
+    }
+
+    get writePermission(): boolean {
+      return this.props.resourceCard.permission === 'OWNER'
+        || this.props.resourceCard.permission === 'WRITER';
+    }
+
+    get displayName(): string {
+      return this.props.resourceCard.notebook.name.replace(/\.ipynb$/, '');
+    }
+
+    get displayDate(): string {
+      if (!this.props.resourceCard.modifiedTime) {
+        return '';
+      }
+
+      const date = new Date(this.props.resourceCard.modifiedTime);
+      // datetime formatting to slice off weekday from readable date string
+      return date.toDateString().split(' ').slice(1).join(' ');
+    }
+
+    resourceUrl(jupyterLab = false): string {
+      const {workspaceNamespace, workspaceFirecloudName, notebook} =
+        this.props.resourceCard;
+      const workspacePrefix = `/workspaces/${workspaceNamespace}/${workspaceFirecloudName}`;
+
+      const queryParams = new URLSearchParams([
+        ['playgroundMode', 'false'],
+        ['jupyterLabMode', String(jupyterLab)]
+      ]);
+
+      if (this.readOnly) {
+        queryParams.set('jupyterLabMode', 'true');
+      }
+      return `${workspacePrefix}/notebooks/preview/${encodeURIComponent(notebook.name)}`;
+    }
+
+    async receiveNotebookRename(newName) {
+      const {resourceCard} = this.props;
+      try {
+        await workspacesApi().renameNotebook(
+          resourceCard.workspaceNamespace,
+          resourceCard.workspaceFirecloudName,
+          {
+            name: resourceCard.notebook.name,
+            newName: this.fullNotebookName(newName)
+          });
+      } catch (error) {
+        console.error(error); // TODO: better error handling
+      } finally {
+        this.setState({renaming: false});
+        this.props.onUpdate();
+      }
+    }
+
+    async deleteNotebook(onComplete: Function) {
+      workspacesApi().deleteNotebook(
+        this.props.resourceCard.workspaceNamespace,
+        this.props.resourceCard.workspaceFirecloudName,
+        this.props.resourceCard.notebook.name)
+        .then(() => {
+          onComplete();
+          this.props.onUpdate();
+        });
+    }
+
+    async duplicateNotebook() {
+      workspacesApi().cloneNotebook(
+        this.props.resourceCard.workspaceNamespace,
+        this.props.resourceCard.workspaceFirecloudName,
+        this.props.resourceCard.notebook.name)
+        .then(() => {
+          this.props.onUpdate();
+        }).catch(e => {
+          this.props.onDuplicateResource(false);
```

Comment:
What exactly is the purpose of this callback? It's only ever called with a 'false' argument here. Should we call it with a 'true' value at L138 before making the API call?

Actually... it seems like this may be redundant, as we already call the onUpdate callback to cause the notebook list to reload altogether. Is there some special handling that the notebook-list needs to do when a notebook is duplicated?

Reply:
good catch. It got copied over while I was refactoring but its no longer necessary.